src/basic/hashmap.c

```diff
@@ -2005,8 +2016,16 @@ int set_strjoin(Set *s, const char *separator, char **ret) {
                 len += l;
                 first = false;
         }
-        if (str)
-                str[len] = '\0';
+
+        if (wrap_with_separator) {
+                if (!GREEDY_REALLOC(str, allocated, len + separator_len + 1))
```

Comment:
It'd be nicer to include this extra len in the GREEDY_REALLOC above. The code would be shorter and we'd avoid potentially one more realloc.